### ISSUE TITLE
Question about properties of LLaMa embeddings

### ISSUE THREAD

**Issue by horenbergerb** [skeskinen/llama-lite#2]
Hi, please feel free to close this question as it doesn't strictly pertain to your repo. I just hoped you might be familiar enough with LLaMa embeddings to answer it.

I'm working on a use case where I want to calculate similarity scores among LLaMa embeddings. The similarity scores are just euclidean distances normalized to be between 0 and 1

For OpenAI embeddings, normalizing the distances is easy because the embeddings have unit norm, which puts a convenient maximum and minimum on the distances. Normalizing is done via (1.0 - distance / math.sqrt(2)).

I'm currently using Vicuna 13B, and what I'm seeing is that the embeddings have dimension 5120, and the norm of the embeddings varied in the range of ~88-93.

tldr: Do you know if there's any restrictions on LLaMa embeddings, such as bounds on what the norm can be, etc? Any properties that would help me normalize the distances between embeddings?

Thanks!

**Comment by skeskinen** [comment]
Hi, I'm not sure I can help with this. I'm more of a programmer and most of the ML math goes over my head.
But maybe you could try taking the embeddings from a different point in the llama.cpp and see if it works better for you:

```
    // norm
    {
        inpL = ggml_rms_norm(ctx0, inpL);

        // inpL = norm*inpL
        inpL = ggml_mul(ctx0,
                    ggml_repeat(ctx0, model.norm, inpL),
                    inpL);

        embeddings = inpL;
    }
```
You could take a look at the embeddings after the rms_norm and before multiplication with the model norm layer?

**Comment by horenbergerb** [comment]
Okay, thanks for the tip! Much appreciated!
